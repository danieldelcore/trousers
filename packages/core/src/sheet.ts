const createStyleElement = (attributeId: string) => {
    const element = document.createElement('style');
    element.setAttribute(attributeId, '');
    element.setAttribute('type', 'text/css');
    element.appendChild(document.createTextNode(''));
    return element;
};

const getElement = (targetElement: HTMLElement, attributeId: string) =>
    targetElement.querySelector<HTMLStyleElement>(`style[${attributeId}]`);

const getSheet = (tag: HTMLStyleElement): CSSStyleSheet => {
    if (tag.sheet) {
        // @ts-ignore
        return tag.sheet;
    }

    for (let i = 0; i < document.styleSheets.length; i++) {
        if (document.styleSheets[i].ownerNode === tag) {
            // @ts-ignore
            return document.styleSheets[i];
        }
    }

    throw Error('Unable to get StyleSheet');
};

const sheet = (targetEl: HTMLElement, attributeId: string) => {
    let styleEl = getElement(targetEl, attributeId);
    const styleMap = new Map();

    if (styleEl === null) {
        styleEl = createStyleElement(attributeId);
        targetEl.appendChild(styleEl);
    }

    const clear = () => styleEl!.remove();

    const has = (id: string) => styleMap.has(id);

    const mount = (id: string, styles: string, isGlobal?: boolean) => {
        styleMap.set(id, '');

        try {
            const sheet = getSheet(styleEl!);
            sheet.insertRule(styles, isGlobal ? 0 : sheet.cssRules.length);

            return;
        } catch (error) {
            console.warn(`Trousers: unable to insert rule: ${styles}`, error);
        }
    };

    return {
        has,
        clear,
        mount,
    };
};

export default sheet;
