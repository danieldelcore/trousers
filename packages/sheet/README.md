# @trousers/sheet

`@trousers/sheet` is responsible for managing the styles that are mounted to the page.

## Installation

`yarn add @trousers/sheet` or `npm install @trousers/sheet`

## API

### `sheet()`

Is the factory function which exposes the sheet API.

**Arguments:**

-   `targetEl` (`HTMLElement`): Element which will have style tags mounted to
-   `attributeId` (`string`): The attribute which Trousers will use to identify the style tag

**Returns:**

-   `has`: (`(id: string) => boolean`): Will return true if id has been mounted to the page
-   `mount`: (`(id: string, styles: string, isGlobal?: boolean => void`): Will mount provided styles to the page. If `isGlobal` is provided style will be mounted globally.
-   `unmount`: (`(id: string) => void`): unmounts all styles associated with the provided `id`

**Example:**

```js
import Sheet from '@trousers/sheet';

const sheet = Sheet(head, 'i-love-trousers');
```

### `sheet().has()`

Will return true if id has been mounted to the page.

**Arguments:**

-   `id` (`string`): The id of the style you're checking for

**Returns:**

-   `boolean`: The status of the style

**Example:**

```js
import Sheet from '@trousers/sheet';

const sheet = Sheet(head, 'i-love-trousers');

sheet.has('bar'); // returns false;
```

### `sheet().mount()`

Will mount provided styles to the page.

**Arguments:**

-   `id` (`string`): unique identifier for the provided style
-   `styles` (`string`): style string
-   `isGlobal` (`boolean`)?: The styles should be treated as global or not

**Returns:**

-   `void`

**Example:**

```js
import Sheet from '@trousers/sheet';

const sheet = Sheet(head, 'i-love-trousers');

sheet.mount('1', '.foo{background-color:red;}');
```

### `sheet().unmount()`

Flushes all styles tracked by the provides `id`

**Arguments:**

-   `id` (`string`): unique identifier for the provided style

**Returns:**

-   `void`

**Example:**

```js
import Sheet from '@trousers/sheet';

const sheet = Sheet(head, 'i-love-trousers');

sheet.mount('1', '.foo{background-color:red;}');
sheet.unmount('1');
```
