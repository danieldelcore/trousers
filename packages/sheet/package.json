{
    "name": "@trousers/sheet",
    "version": "4.0.0-alpha2",
    "main": "dist/trousers-sheet.cjs.js",
    "module": "dist/trousers-sheet.esm.js",
    "types": "dist/index.d.ts",
    "license": "MIT",
    "dependencies": {},
    "devDependencies": {},
    "repository": "https://github.com/danieldelcore/trousers/tree/master/packages/sheet"
}
