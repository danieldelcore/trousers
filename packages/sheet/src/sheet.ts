const createStyleElement = (attributeId: string) => {
    const element = document.createElement('style');
    element.setAttribute(attributeId, '');
    element.setAttribute('type', 'text/css');
    element.appendChild(document.createTextNode(''));
    return element;
};

const getElement = (targetEl: HTMLElement, attributeId: string) =>
    targetEl.querySelector<HTMLStyleElement>(`style[${attributeId}]`);

const getSheet = (targetEl: HTMLStyleElement) => {
    if (targetEl.sheet) return targetEl.sheet;

    for (let i = 0; i < document.styleSheets.length; i++) {
        if (document.styleSheets[i].ownerNode === targetEl) {
            return document.styleSheets[i];
        }
    }

    throw Error('Unable to get StyleSheet');
};

const sheet = (targetEl: HTMLElement, attributeId: string) => {
    let styleEl = getElement(targetEl, attributeId);
    const styleMap = new Map<string, number>();

    if (styleEl === null) {
        styleEl = createStyleElement(attributeId);
        targetEl.appendChild(styleEl);
    }

    const pushOrder = () =>
        styleMap.forEach((value, key, map) => map.set(key, (value += 1)));

    const popOrder = (popIndex: number) => {
        styleMap.forEach((value, key, map) => {
            const index = Array.from(map.keys()).indexOf(key);

            if (popIndex <= index) {
                map.set(key, (value -= 1));
            }
        });
    };

    const has = (id: string) => styleMap.has(id);

    const mount = (id: string, styles: string, isGlobal?: boolean) => {
        if (has(id)) return;

        try {
            const activeSheet = getSheet(styleEl!);
            // TODO: this will break because the ids change over time as new elements like globals are pushed in
            const index = isGlobal ? 0 : activeSheet.cssRules.length;
            activeSheet.insertRule(styles, index);

            if (isGlobal) pushOrder();

            styleMap.set(id, index);
        } catch (error) {
            console.warn(`Trousers - unable to insert rule: ${styles}`, error);
        }
    };

    const unmount = (id: string) => {
        if (!has(id)) return;

        const index = styleMap.get(id) as number;
        const activeSheet = getSheet(styleEl!);
        activeSheet.deleteRule(index);
        styleMap.delete(id);
        popOrder(index);
    };

    return {
        has,
        mount,
        unmount,
    };
};

export default sheet;
