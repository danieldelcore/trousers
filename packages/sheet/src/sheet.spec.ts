import sheet from './sheet';

declare global {
    namespace jest {
        interface Matchers<R> {
            toHaveMountedStyles(key: string, style: object): R;
        }
    }
}

const objectContaining = (
    obj: Record<string, any>,
    subObj: Record<string, any>,
) =>
    Object.keys(subObj).reduce((accum, key) => {
        if (accum === false || !obj.hasOwnProperty(key)) return false;
        return obj[key] === subObj[key];
    }, true);

const styleToString = (key: string, style: object) => {
    const styles = Object.entries(style).reduce(
        (accum, [rule, property]) => `${accum}${rule}:${property};`,
        '',
    );

    return `${key}{${styles}}`;
};

expect.extend({
    toHaveMountedStyles(cssRuleList, selectorText, rules) {
        const pass =
            cssRuleList.selectorText === selectorText &&
            objectContaining(cssRuleList.style, rules);

        const message = pass
            ? () =>
                  `Expected: not ${selectorText} ${this.utils.printExpected(
                      cssRuleList,
                  )}\n` +
                  `Received: ${
                      cssRuleList.selectorText
                  }${this.utils.printReceived(rules)}`
            : () =>
                  `Expected: ${selectorText} ${this.utils.printExpected(
                      cssRuleList,
                  )}\n` +
                  `Received: ${
                      cssRuleList.selectorText
                  }${this.utils.printReceived(rules)}`;

        return {
            pass,
            message,
        };
    },
});

describe('Sheet', () => {
    let activeSheet: ReturnType<typeof sheet>;
    let headEl = window.document.head;
    let attributeId = 'data-testing-trousers';
    let styleSheet: CSSStyleSheet;

    beforeEach(() => {
        activeSheet = sheet(headEl, attributeId);
        styleSheet = document.styleSheets[0];
    });

    afterEach(() => {
        window.document.head.innerHTML = '';
    });

    it('can mount a style tag', () => {
        sheet(headEl, 'data-testing-secondary');
        expect(headEl.innerHTML).toMatchSnapshot();
    });

    it('detects an existing style', () => {
        const key = '.foo';
        activeSheet.mount(key, styleToString(key, { color: 'red' }));

        expect(activeSheet.has(key)).toEqual(true);
    });

    it('detects a non existing style', () => {
        expect(activeSheet.has('123')).toBe(false);
    });

    it('mounts a new style', () => {
        const key = '.foo';
        const style = { color: 'red' };
        const styleRule = styleToString(key, style);

        activeSheet.mount(key, styleRule);

        expect(activeSheet.has(key)).toEqual(true);
        expect(headEl.innerHTML).toMatchSnapshot();
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style);
    });

    it('mounts style with multiple properties', () => {
        const key = '.foo';
        const style = { color: 'blue', backgroundColor: 'red' };
        const styleRule = styleToString(key, style);

        activeSheet.mount(key, styleRule);

        expect(activeSheet.has(key)).toEqual(true);
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style);
    });

    it('mounts multiple styles', () => {
        const key1 = '.blue';
        const key2 = '.red';
        const style1 = { color: 'blue' };
        const style2 = { color: 'red' };

        activeSheet.mount(key1, styleToString(key1, style1));
        activeSheet.mount(key2, styleToString(key2, style2));

        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key1, style1);
        expect(styleSheet.cssRules[1]).toHaveMountedStyles(key2, style2);
    });

    it('mounts a global style', () => {
        const key = '.1';
        const style = { color: 'red' };

        activeSheet.mount(key, styleToString(key, style));

        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style);
    });

    it('mounts a global, with pre-existing style tag', () => {
        const key1 = '.blue';
        const key2 = 'body';
        const style1 = { color: 'blue' };
        const style2 = { color: 'red' };

        activeSheet.mount(key1, styleToString(key1, style1));
        activeSheet.mount(key2, styleToString(key2, style2), true);

        expect(styleSheet.cssRules[1]).toHaveMountedStyles(key1, style1);
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key2, style2);
    });

    it('registering an existing key does not update the style', () => {
        const key = '.blue';
        const style1 = { color: 'blue' };
        const style2 = { color: 'red' };

        activeSheet.mount(key, styleToString(key, style1));
        activeSheet.mount(key, styleToString(key, style2));

        expect(styleSheet.cssRules.length).toEqual(1);
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style1);
    });

    it('unmounts style as the only mounted style', () => {
        const key = '.foo';
        const style = { color: 'red' };
        const styleRule = styleToString(key, style);

        activeSheet.mount(key, styleRule);

        expect(activeSheet.has(key)).toEqual(true);
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style);

        activeSheet.unmount(key);
    });

    it('unmounts style as one of many mounted styles', () => {
        const key = '.foo';
        const style = {
            color: 'red',
        };

        activeSheet.mount('.bar', '.bar{color: blue;}');
        activeSheet.mount(key, styleToString(key, style));
        activeSheet.mount('.baz', '.baz{color: green;}');

        expect(activeSheet.has(key)).toEqual(true);
        expect(styleSheet.cssRules.length).toEqual(3);
        expect(styleSheet.cssRules[1]).toHaveMountedStyles(key, style);

        activeSheet.unmount(key);

        expect(activeSheet.has(key)).toEqual(false);
        expect(styleSheet.cssRules.length).toEqual(2);
        expect(styleSheet.cssRules[0]).not.toHaveMountedStyles(key, style);
        expect(styleSheet.cssRules[1]).not.toHaveMountedStyles(key, style);
    });

    it('mounts then unmount all styles in order', () => {
        const key = '.foo';
        const style = {
            color: 'red',
        };
        const key2 = '.bar';
        const style2 = {
            color: 'blue',
        };
        const key3 = '.baz';
        const style3 = {
            color: 'green',
        };

        activeSheet.mount(key, styleToString(key, style));
        activeSheet.mount(key2, styleToString(key2, style2));
        activeSheet.mount(key3, styleToString(key3, style3));

        expect(styleSheet.cssRules.length).toEqual(3);
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style);
        expect(styleSheet.cssRules[1]).toHaveMountedStyles(key2, style2);
        expect(styleSheet.cssRules[2]).toHaveMountedStyles(key3, style3);

        activeSheet.unmount(key);
        expect(activeSheet.has(key)).toEqual(false);
        expect(styleSheet.cssRules.length).toEqual(2);
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key2, style2);
        expect(styleSheet.cssRules[1]).toHaveMountedStyles(key3, style3);

        activeSheet.unmount(key2);
        expect(activeSheet.has(key2)).toEqual(false);
        expect(styleSheet.cssRules.length).toEqual(1);

        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key3, style3);

        activeSheet.unmount(key3);
        expect(activeSheet.has(key3)).toEqual(false);
        expect(styleSheet.cssRules.length).toEqual(0);
    });

    it('mounts then unmounts all styles from middle of array', () => {
        const key = '.foo';
        const style = {
            color: 'red',
        };
        const key2 = '.bar';
        const style2 = {
            color: 'blue',
        };
        const key3 = '.baz';
        const style3 = {
            color: 'green',
        };

        activeSheet.mount(key, styleToString(key, style));
        activeSheet.mount(key2, styleToString(key2, style2));
        activeSheet.mount(key3, styleToString(key3, style3));

        expect(styleSheet.cssRules.length).toEqual(3);
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style);
        expect(styleSheet.cssRules[1]).toHaveMountedStyles(key2, style2);
        expect(styleSheet.cssRules[2]).toHaveMountedStyles(key3, style3);

        activeSheet.unmount(key2);
        expect(activeSheet.has(key2)).toEqual(false);
        expect(styleSheet.cssRules.length).toEqual(2);
        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style);
        expect(styleSheet.cssRules[1]).toHaveMountedStyles(key3, style3);

        activeSheet.unmount(key);
        expect(activeSheet.has(key)).toEqual(false);
        expect(styleSheet.cssRules.length).toEqual(1);

        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key3, style3);

        activeSheet.unmount(key3);
        expect(activeSheet.has(key3)).toEqual(false);
        expect(styleSheet.cssRules.length).toEqual(0);
    });

    it('unmounts a global style', () => {
        const key = ':root';
        const style = { color: 'red' };

        activeSheet.mount(key, styleToString(key, style), true);

        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key, style);
    });

    it('unmounts a global style as one of many mounted global styles', () => {
        const key = ':root';
        const style = { color: 'red' };
        const key2 = 'body';
        const style2 = { color: 'blue' };
        const key3 = '*';
        const style3 = { color: 'green' };

        activeSheet.mount(key, styleToString(key, style), true);
        activeSheet.mount(key2, styleToString(key2, style2), true);
        activeSheet.mount(key3, styleToString(key3, style3), true);

        expect(activeSheet.has(key)).toEqual(true);
        expect(styleSheet.cssRules.length).toEqual(3);
        expect(styleSheet.cssRules[2]).toHaveMountedStyles(key, style);

        activeSheet.unmount(key);

        expect(styleSheet.cssRules.length).toEqual(2);

        expect(styleSheet.cssRules[0]).toHaveMountedStyles(key3, style3);
        expect(styleSheet.cssRules[1]).toHaveMountedStyles(key2, style2);
    });
});
