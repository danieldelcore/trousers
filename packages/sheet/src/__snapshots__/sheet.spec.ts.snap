// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sheet can mount a style tag 1`] = `
<style data-testing-trousers="" type="text/css"></style>
<style data-testing-secondary="" type="text/css"></style>
`;

exports[`Sheet mounts a new style 1`] = `<style data-testing-trousers="" type="text/css"></style>`;
